### PR TITLE
Implemented a 5-star average review system.

### PR DESCRIPTION
**Pull request by Shromm-Gaind**
In `BookService`, I added methods to calculate the average rating and retrieve the number of ratings for a book.

In `BooksController`, I updated the `updateBooks` method to display the average rating and number of ratings for each book using the new methods in `BookService`.

I had some fun while writing this code, so I fucked around in the `BookService` class—deny it if you don't see my logic:

- I implemented the Bill Pugh Singleton pattern, as it ensures thread safety and lazy initialization without requiring synchronization blocks. 

- I implemented caching for book reviews by using a simple in-memory cache with `ConcurrentHashMap` to store and retrieve review data for books. so this should minimize repeated database access for frequently accessed books.

### CODE REVIEW

**Comment by Shromm-Gaind** [comment]
## Additional Fixes and Improvements:

### Review Submission and Update:

- In `ReviewController`, I added logic to handle review submissions. If a user already has a review for a book, it will be updated; otherwise, a new review will be added. This is handled via `bookService.addOrUpdateReview()`.

### Mock DAO (`ReviewMAO`) Updates:

- Implemented the `saveOrUpdateReview()` method in `ReviewMAO` to align with the interface updates. This method checks if a review for the given user and book exists and updates it if found, or adds a new review if not.
- Ensured that the `ReviewMAO` class is now fully compliant with the updated `IReviewAO` interface, addressing the abstract method implementation error.

File: src/main/java/com/bookclub/controller/BooksController.java [hunk 1]
```diff
@@ -30,12 +33,13 @@ public class BooksController {
...
     public BooksController() {
         bookAO = new BookDAO();
+        bookService = BookService.getInstance();
     }
 
     @FXML
     public void initialize() {
-        // Set nav bar button colour
-        Button booksButton = (Button)navBar.lookup("#booksButton");
+        // Set nav bar button color
```
Comment (jmccook96) [inline src/main/java/com/bookclub/controller/BooksController.java:41]: American GPT detected 😆

File: src/main/java/com/bookclub/service/BookService.java [hunk 2]
```diff
@@ -1,28 +1,40 @@
...
 
     private BookService() {
         bookAO = new BookDAO();
+        reviewAO = new ReviewDAO();
         selectedBook = new SimpleObjectProperty<>();
+        reviewCache = new ConcurrentHashMap<>();
+    }
+
+    private static class BookServiceHolder {
+        private static final BookService INSTANCE = new BookService();
```
Comment (jmccook96) [inline src/main/java/com/bookclub/service/BookService.java:30]: Woah that's a very formal way of doing a singleton... If we go down this rabbit hole for configuration we should update the other configurations/singleton services to match. Don't mind either way. I'm more on the like it side of the fence.

File: src/main/java/com/bookclub/service/BookService.java [hunk 3]
```diff
@@ -35,4 +47,45 @@ public void setSelectedBook(Book book) {
...
+        reviewCache.remove(book);
+    }
+
+    // Add or update review and invalidate the cache
+    public void addOrUpdateReview(Review review, Book book) {
+        reviewAO.saveOrUpdateReview(review);
+        clearReviewCache(book);
+    }
+
+    public static class ReviewData {
```
Comment (jmccook96) [inline src/main/java/com/bookclub/service/BookService.java:74]: Huh today I learnt about static classes. Nice

**Review by jmccook96 — APPROVED** [review]

**Pull request merged**